qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+pulse programming. On the contrary, it removes parameter table object and associated logic
+from each object, yielding smaller object creation cost and higher performance
+as the data amount scales.
+
+Note that we don't need to write any parameter management logic for each object,
+and thus this parameter framework gives greater scalability to the pulse module.
+"""
+from copy import deepcopy, copy
+from typing import List, Dict, Set, Any, Union
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parameterexpression import ParameterExpression, ParameterValueType
+from qiskit.pulse import instructions, channels
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.library import ParametricPulse, Waveform
+from qiskit.pulse.schedule import Schedule, ScheduleBlock
+from qiskit.pulse.transforms.alignments import AlignmentKind
+from qiskit.pulse.utils import format_parameter_value
+
+
+class NodeVisitor:
+    """A node visitor base class that walks instruction data in various pulse programs
+    and calls visitor functions for every node found.
+
+    Though this class implementation is based on Python AST, each node doesn't have
+    dedicated node class due to lack of abstract syntax tree for pulse programs in Qiskit.
+    Instead of parsing pulse programs, this visitor class finds the associated visitor function
+    based on class name of the instruction node, i.e. ``Play``, ``Call``, etc...
+    The `.visit` method recursively checks superclass of given node since some parametrized
+    components such as ``DriveChannel`` may share common superclass with other subclasses.
+    In this example, we can just define ``visit_Channel`` method instead of defining
+    the same visitor function for every subclasses.
+
+    Some instructions may have special logic or data structure to store parameter objects,
+    and visitor functions for these nodes should be individually defined.
+
+    Because pulse programs can be nested into another pulse program,
+    the visitor function should be able to recursively call proper visitor functions.
+    If visitor function is not defined for a given node, ``generic_visit``
+    method is called. Usually this method is provided for operating on object
+    defined outside of the Qiskit Pulse module.
+    """
+    def visit(self, node: Any):
+        """Visit a node."""
+        visitor = self._get_visitor(node.__class__)
+        return visitor(node)
+
+    def _get_visitor(self, node_class):
+        """A helper function to recursively investigate superclass visitor method."""
+        if node_class == object:
+            return self.generic_visit
+
+        try:
+            return getattr(self, f'visit_{node_class.__name__}')
+        except AttributeError:
+            # check super class
+            return self._get_visitor(node_class.__base__)
+
+    def visit_ScheduleBlock(self, node: ScheduleBlock):
```

Comment:
I'm not a fan of this casing style. I believe it should be `visit_schedule_block`, etc.

Reply:
This probably means that you need an explicit lookup table between object names and functions, which I think is ok. Or a decorator of some sort which I believe is used in `converters/pulse_instructions.py`

Reply:
This logic is actually based on Python AST implementation. This is a lightweight logic to map instance to visitor function. We don't need to maintain a separate mapper when new node type is added. But this logic may cause a problem when we start to use auto code formatter such as [black](https://github.com/psf/black).

This is already used in the parameter parser.
https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/pulse/parser.py